vowpalwabbit/automl.cc

```
@@ -0,0 +1,743 @@
+    ec[0]->pred.a_s = std::move(data.champ_a_s);
+  }
+
+  // extra: assert again just like at the top
+  assert(data.all->interactions.empty() == true);
+  assert(ec[0]->interactions == nullptr);
+}
+
+void persist(automl& data, metric_sink& metrics) { data.cm.persist(metrics); }
+
+void finish_example(vw& all, automl& data, multi_ex& ec)
+{
+  {
+    size_t champ_stride = data.cm.current_champ;
+    for (example* ex : ec) { data.cm.apply_config(ex, champ_stride); }
+
+    auto restore_guard = VW::scope_exit([&data, &ec, &champ_stride] {
+      for (example* ex : ec) { data.cm.revert_config(ex); }
+    });
+
+    data.adf_learner->print_example(all, ec);
+  }
+
+  VW::finish_example(all, ec);
+}
+
+void save_load_aml(automl& d, io_buf& model_file, bool read, bool text)
+{
+  if (model_file.num_files() == 0) { return; }
+  if (!read || d.model_file_version >= VERSION_FILE_WITH_AUTOML) { d.cm.save_load(model_file, read, text); }
+}
+
+VW::LEARNER::base_learner* automl_setup(VW::setup_base_i& stack_builder)
+{
+  options_i& options = *stack_builder.get_options();
+  vw& all = *stack_builder.get_all_pointer();
+
+  size_t budget;
+  size_t max_live_configs;
+
+  option_group_definition new_options("Debug: automl reduction");
+  new_options
+      .add(make_option("automl", max_live_configs)
+               .necessary()
+               .keep()
+               .default_value(3)
+               .help("set number of live configs"))
+      .add(make_option("budget", budget).keep().default_value(10).help("set initial budget for automl interactions"));
+
+  if (!options.add_parse_and_check_necessary(new_options)) return nullptr;
+
+  auto data = VW::make_unique<automl>(budget, all.model_file_ver, max_live_configs);
+  // all is not needed but good to have for testing purposes
+  data->all = &all;
+  assert(max_live_configs <= MAX_CONFIGS);
+
+  // override and clear all the global interactions
+  // see parser.cc line 740
+  all.interactions.clear();
```

You should raise an error if this is non-empty as that would be surprising behavior

---

I think at this point it already has the parsed interactions of the cmd line populated here, but now you are making me doubt.